### ISSUE TITLE
Cross platform cookie file format

### ISSUE THREAD

**Issue by macf0x** [costastf/locationsharinglib#66]
Currently the cookie file is encoded with pythons pickle dump function, requiring a Python env. to generate.

If the library supported simple text file or other encoding, it would make the use of this great library more accessible to those less familiar with python. Much like wget/curl operate.

i.e.
```
wget -x --load-cookies cookies.txt "https://www.google.com/maps/@40.7484986,-73.9857129,15z?hl=en"
curl --cookie cookies.txt "https://www.google.com/maps/@40.7484986,-73.9857129,15z?hl=en"
```
This is particularly the case with apps like Home Assistant (hass.io), where python env. is not exposed to run the mapscookiegettercli

Where the library supported a plain text file, an extensions such as [Export Cookies txt](https://chrome.google.com/webstore/detail/cookiestxt/njabckikapfpffapmjgojcnbfjonfjfg) could be used.

Alternatively other scripting languages could leverage the Selenium Webdriver to extract cookies.

```powershell
#Credits https://tech.mavericksevmont.com/blog/powershell-selenium-automate-web-browser-interactions-part-i/
#Credits https://github.com/costastf/mapscookiegettercli

$MAPS_LOGIN = 'https://accounts.google.com/signin/v2/identifier?'`
              +'hl=en&'`
              +'passive=true&'`
              +'continue=https%3A%2F%2Fwww.google.com%2Fmaps%2F%4040.7484986%2C-73.9857129%2C15z%3Fhl%3Den&'`
              +'service=local&'`
              +'flowName=GlifWebSignIn&'`
              +'flowEntry=ServiceLogin'

$LOGGED_IN_HEURISTIC = 'Find local businesses, view maps and get driving directions in Google Maps.'


$scriptPath = split-path -parent $MyInvocation.MyCommand.Definition
if($env:PATH -notlike "*$scriptPath*") { $env:PATH += ";$scriptPath\" }

Add-Type -Path "$scriptPath\WebDriver.dll"
$ChromeDriver = New-Object OpenQA.Selenium.Chrome.ChromeDriver # Creates an instance of this class to control Selenium and stores it in an easy to handle variable

$ChromeDriver.Navigate().GoToURL($MAPS_LOGIN)
while ($ChromeDriver.PageSource -notlike "*$LOGGED_IN_HEURISTIC*" )
{ 
    Sleep -Milliseconds 500
}
$ChromeDriver.Manage().Cookies.AllCookies.GetEnumerator()

#Code to store file in json/plaintext
```

**Comment by costastf** [comment]
Makes a lot of sense, I will definitely look into it. Sadly my time is very very limited so it will take some time.

**Comment by macf0x** [comment]
Glad you think so. Look forward to seeing it.

**Comment by costastf** [comment]
#69 mentions that the cookie getter project is blocked by google so I this issue is becoming a bit higher priority. I will try to implement the feature during the next 10 days if possible. Just FYI.

**Comment by costastf** [comment]
I just pushed a WIP branch with a simple implementation of a text based cookie support https://github.com/costastf/locationsharinglib/tree/text-based-cookies. Sadly cannot test currently because my ip is blocked by google :) . Just a heads up.

**Comment by costastf** [comment]
https://github.com/costastf/locationsharinglib/tree/text-based-cookies has been tested with firefox with this addon https://github.com/rotemdan/ExportCookies (Prefix HttpOnly cookies has to be unchecked) and with chrome with this addon https://chrome.google.com/webstore/detail/cookiestxt/njabckikapfpffapmjgojcnbfjonfjfg and they both work fine. If you want to give that a spin and report back any possible issues that would be great!

**Comment by costastf** [comment]
release 4.1.0 implements text based cookies. Thanks a lot for this great idea @macf0x and apologies for the long implementation time!

**Comment by macf0x** [comment]
Awesome. Sorry been on leave. Thus playing catch up. I'll give this a test.

**Comment by macf0x** [comment]
@costastf 

Tested and looks to be working.

Only trick is to hit the right URL to get the full complement of cookies.

URL in your cli code seems to do the trick.

https://accounts.google.com/signin/v2/identifier?hl=en&passive=true&continue=https%3A%2F%2Fwww.google.com%2Fmaps%2F%4040.7484986%2C-73.9857129%2C15z%3Fhl%3Den&service=local&flowName=GlifWebSignIn&flowEntry=ServiceLogin

Hitting maps.google.com.au doesn't seems to generate the right cookies. but maps.google,com does.



```
(google_maps) macfox@ubuntu:~/locationsharinglib$ python3 lsltest.py
Full name        :xxxx yyyy
Nickname         :xxxxx
Current location :xxxxxxxxxxxxxxxxxxxxxxxxxxxx
Latitude         :-xx.0499019
Longitude        :xxx.5852517
Datetime         :2019-09-04 14:36:51.846000+00:00
Charging         :False
Battery %        :40
Accuracy         :14
Full name        :xxx@xxxx.com
Nickname         :xxx@xxxxxx.com
Current location :Xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
Latitude         :-xx.049926
Longitude        :xxxx.5852543
Datetime         :2019-09-04 14:14:15.438000+00:00
Charging         :False
Battery %        :None
Accuracy         :14
```

**Comment by costastf** [comment]
Do you get an error that the cookies are invalid when you visit the au domain? I am actually checking the validity of the cookies by trying to access myaccount.google.com . It might be that the au domain cookies only give access to au myaccount domain. The problem of course is that it's too complex to try to guess the users location and try to use that domain and even if that is implemented there is no way to make sure that the user actually used the local domain. So I think the only way out of this is to actually explicitly mention in the documentation that the domain should be the .com one. Thoughts?

**Comment by macf0x** [comment]
I concur. I'd specifically say you need to visit the .com URL. No need to complicate things. Just thought I'd mention that in case it comes up.

BTW. I was at Google SYD last week and spoke to the Maps team. I specifically raised the lack of API access for location sharing with the team and PMs. I explained our use case. Not sure it will result in much, as this topic was very left field to the feedback/requests they were seeking. But the PM now has visibility.

I'll follow up with them in a few weeks.

**Comment by costastf** [comment]
I would guess that, that is probably why the mapscookiegettercli project
got blocked :) The intention is great, but somehow I don't think it will
work to our benefit...

**Comment by macf0x** [comment]
I don't think the cli app was targeted due to its intent. Google have various systems that pickup anomalies, to combat abuse. Anything that looks out of wack will get that kind of attention.

I will be following up with them regardless. Also the feedback option is worth using. The PMs do actually look at these, so we should be encouraging users to request this, with a link to your project.

**Comment by costastf** [comment]
Yeah, of course, it would be the ideal situation. Thanks for running with
this, on behalf of all users of location sharing :)

**Comment by macf0x** [comment]
Thank you for this great library.

**Comment by Bijan-A** [comment]
Is there a solution that does not utilize dataclasses module? I am on Python 3.5 and cannot upgrade to 3.6+ on my machine so I receive syntax error when it gets to "domain: str"

**Comment by costastf** [comment]
Hey @Abolfazl , sorry for taking so long to answer,I was on vacation. You could replace the data classes with a name tuple and everything should work the same. I don't want to extend the compatibility to older python versions, but if you have trouble doing it yourself I could help out once of the upcoming days. I will not be making an official release with named tuples though.

**Comment by lolomin** [comment]
Hi,

Just found this issue which is precisely the one I was having "Couldn't sign you in" when trying to use maps-cookie-getter ...

Sorry to bother but I'm in no way a Python developer, just a simple sysadmin which have some notions on PHP or Bash but in no way in Python and I already passed some great time configuring the previous last version of your library searching for code example on how to retrieve the cookie and how to retrieve informations about location but now after reading all this issue's comments, I'm, again, still lost : 

On Sept 1st 2019 you said "release 4.1.0 implements text based cookies. Thanks a lot for this great idea @macf0x and apologies for the long implementation time!"

Ok Great and thks :) but how does it work ? How to retrieve the cookie by now please ? I have tried with the Chrome extension which leads me to download a cookie.txt file for the active Google Maps tab but after trying to give this cookie.txt file to locationsharinglib I have the following error :

TypeError: __init__() got multiple values for argument 'cookies_file'

Content of googlemap.py is :

from locationsharinglib import Service
service = Service("xxxxxxxx@gmail.com", "passwdxxxxx", cookies_file="cookie.txt")
for person in service.get_all_people():
    print(person)

Could not find any documentation on usage of the new cookie getter procedure ? Could you help please ?

Regards,

Laurent

**Comment by costastf** [comment]
Hi lolomin,
no need to apologize, been there, done that I know how it is to be lost :)
So the v4 major version changed the way the library works internally and it
does not try to do any type of authentication any more, instead it depends
on the user having done the authentication through a browser and the extracting the cookies to a cookie file through an extension. If you look at the usage page of the documentation of the project it is documented how it is supposed to work. https://locationsharinglib.readthedocs.io/en/latest/usage.html 
Just for clarity I will try to expand that since it is quite small.

from locationsharinglib import Service
cookies_file = 'FILE_CREATED_BY_MAPSCOOKIEGETTERCLI_AUTHENTICATION_PROCESS'
google_email = 'username@gmail.com'
service = Service(cookies_file=cookies_file, authenticating_account=google_email)
for person in service.get_all_people():
    print(person)

where cookies_file should be a path toward the cookie file exported by your browser add on.
This thread should give a lot o details on how that should be set up in the context of Home Assistant. I hope that helps, if not feel free to ping me for clarifications.

**Comment by lolomin** [comment]
Hi costastf,

Thanks for your quick answer.
I have finally succeed to make it work but it seems now unstable as the cookie informations get quickly unusable :
- for a cookie.txt file generated under Windows 10 / Firefox : after a few hours (4 ~ 5 I would say ...) I got no errors but no data is then returned, it seems Google detects too many requests from the IP address and ask for a catcha if I go from a web browser to maps.google.com
- for a cookie.txt file generated from my Linux server / Firefox, it seems to last a few hours more than the one on Windows but after some time I got some errors returned like if the cookie data was expired.

Regards,

Laurent

**Comment by costastf** [comment]
Hi lolomin,
that does not have anything to do with the implementation i am afraid, it
is just Google getting too many requests from the same IP and flagging it.
One thing I could suggest is to make the cookie on a machine with a
different internet ip, like on another network and see how that works for
you.

kind regards
Costas

**Comment by lolomin** [comment]
Hi costastf,

Thanks for your answer.

Tried with a cookie generated from another IP address but failed also a few hours after generated, declared invalid by the lib :: 

service = Service(cookies_file=cookies_file, authenticating_account=google_email)
  File "/opt/locationsharinglib_20200117/locationsharinglib/locationsharinglib.py", line 101, in __init__
    self._session = self._validate_cookie(cookies_file or '')
  File "/opt/locationsharinglib_20200117/locationsharinglib/locationsharinglib.py", line 112, in _validate_cookie
    raise InvalidCookies(message)
locationsharinglib.locationsharinglibexceptions.InvalidCookies: The cookies provided do not provide a valid session, could not reach personal account page.Please create another cookie file and try again.

It seems something has changed IMHO from the cookies generated before (which last 7 months for the same account) and the new cookies file.

To my opinion, it makes the lib useless at this moment as there is no way to regenerate automatically the cookies file after expiration if it only last a few hours, is it me or others people are having the same behavior please?

Maybe something is possible with tools like Puppeteer (https://github.com/puppeteer/puppeteer) ? 

Regards, 

Laurent

**Comment by costastf** [comment]
Hi lolomin,

As far as I know, there are no other reports of such activity. Could it be
that your network traffic patterns changed and that google is
actively flagging you more aggressively?

kind regards
Costas

**Comment by lolomin** [comment]
Don't know exactly what you mean by network traffic patterns ?
But on my side, nothing else has changed other than the version of library and the method to connect to, I can try to export the lib to another server which will have another ougoing IP address and see if I get the same behavior.
I can also try with another Google account.
Will let you know.

**Comment by Bijan-A** [comment]
I am having similar problem as @lolomin. Before, when the session was generated by cookie getter cli, my Google session would never need an update. Now with the cookie file, it seems to expire every couple of days.

**Comment by lolomin** [comment]
Just tested with another Google account from another IP address and so a new cookie file generated with this account on Firefox with ExportCookies add-on, the cookies file was generated at 13:56 it was working well giving all my shared people's position for 2 hours and 12 mins and at 16:08 (04:08 GMT+1 time) it stops working and now the result is constantly this one  :

[root@cygnus locationsharinglib]# python3 00_lolomin_gmap_sharing_infos_grabber.py
[root@cygnus locationsharinglib]# echo $?
0

Content of 00_lolomin_gmap_sharing_infos_grabber.py is : 

from locationsharinglib import Service
cookies_file = '/opt/locationsharinglib/00_lolomin_gmap_cookies-google-com.txt'
google_email = 'xxxxxxxxxxx@gmail.com'
service = Service(cookies_file=cookies_file, authenticating_account=google_email)
for person in service.get_all_people():
    print(person)

Regards,

Laurent

**Comment by costastf** [comment]
Sadly it is really difficult for me to help on this one as that would
require access to your personal information. One thing that is worth doing
is checking whether the cookies you get with the addon match the ones that
you have on a browser session. Have you tried with the chrome addon
"cookies.txt" ?

**Comment by Bijan-A** [comment]
Yes, I used the cookies.txt extension each time.

**Comment by costastf** [comment]
Would it be, that you did that on a windows box. There have been reports
that that extension does not return full cookies on a windows box. Can you
try on a linux or mac one?

**Comment by lolomin** [comment]
Hi,

sorry, was not checking this thread for 2 days : tests were done on Windows 10 AND Fedora 31 with Google Chrome (cookies.txt extension) and Firefox (export cookies extension) with 2 different google accounts. Results were the same : the cookie information seems to last only a few hours for me :/ 
Can't understand why at the moment but IMO this is almost sure there is something different with the cookies information gathered before v4.x and the one gathered with these extensions.

**Comment by lolomin** [comment]
Hi,

Tried some others combinations OS/Browser/spaced requests to Google with an interval of 1 min rather than 30 seconds like I was doing before but nothing succeeds : the cookie finally expired after a few hours.
So my question is : is there some people using this library here with no problem about cookies expiration since the new cookie export procedure (4.x ?) please ? or maybe no one here still using the library anymore and migrate to other projects ? Thanks !

**Comment by au190** [comment]
Iam using Firefox browser to get the cookies. For me its working, no expiration, more then 1 moths.

**Comment by lolomin** [comment]
Thanks au190 for your reply, could you please describe in details how you did the export please (OS version, Firefox version, procedure ?).
I think I'm missing something lol

**Comment by au190** [comment]
Iam using this : https://github.com/rotemdan/ExportCookies
:)

**Comment by lolomin** [comment]
Hi,

After many, many, many attempts and tries, still not able to make this library work properly more than 24 hours with a one min interval script grabbing location sharing data, would very glad to see/have a complete how to/procedure from someone here which is able to use the library without having to renew the cookie please ?

Thks,

lolo

**Comment by Bijan-A** [comment]
i think one minute is too frequent. What happens if you switch to 3-5min?

**Comment by macf0x** [comment]
Ditto.
AFAIK, Android devices do not update their location that frequently, so checking <5 minutes is probably isn't going to make tracking more responsive.

You maybe hitting some Abuse AI. I would try backing it off.

I have this setup with HA, an it tracks multiple mobiles for months without issues.

What I did find that unless you are a consumer account gSuite Basic a a fixed Cookie timeout, where as under the Business and Enterprise editions, you can adjust the timeout or make infinite (like consumer accounts)

Rob

**Comment by lolomin** [comment]
Hi,

Thks for your answer, what I don't understand is that with the previous way of working of the library, I was grabbing the information every 30 seconds and it was working for months. What is the difference now? I only see one thing, the fact that we were able to login each time with email / password so Google was almost sure that it was a really wanted login process, a low probability to be an hijacked login tentative, maybe with the cookie only, it can be considered by Google to be less secure?
From my experience over the previous month and all the data logged, I can say that devices seems to update their location more frequently ( almost every 30 to 60 seconds as soon at it detects a quick/high movement between the previous location, as soon as the location does not move at all the update gets higher but never more than every 3 mins max, so it seems preferable to have a check /refresh frequency of at least 1 min IMHO.

Thanks for your time, will try to update every 1 minute but with two different people accounts Alternatively so every 2 minutes for the same account and will report what will be the result, hope that the source IP address will not be detected as a problem also...

**Comment by lolomin** [comment]
Hi,

Tried for the second time with a 3 mins interval and cookie lasted for approximately 2 days each time but finally expired, again I can't understand why when logging though the previous method with login/password  it was working for months and how are you using this lib without having cookies expired ... I'm probably missing something. 
Will now try with a 5 mins interval but with 3 mins interval it already clearly makes the detection less quicker for actions based on geolocation on a domotic system. If not working with 5 mins interval, will probably switch to solution like owntracks ...

Regards. 

Laurent

**Comment by costastf** [comment]
Hi Laurent,
I am sorry to hear you are having so much trouble and also i am sorry for
not responding more actively but it has been very very hectic. Frankly
there is not much that can be done so i am answering just to give some
history that might help you with troubleshooting or understanding.

This project has seen three evolution stages. The first on implemented
google authentication directly by doing the authentication dance and
submitting to the web url of the authentication mechanism of google to get
a valid session. That worked quite well, but google has many many
authentication mechanisms all over the world so depending on where you
lived you might not be so lucky and also it did not support any other
setting apart from a simple non MFA account.

The second iteration was with an external process that drove a browser via
selenium which essentially could support any account setting since it was
fully human interactive. The problem with that approach is that selenium is
really easy to identify, and of course there are work arounds for that but
that is a cat and mouse game that is not fun to play against google and I
do not have the time, nor the inclination to do that.

So current iteration depends on a person using a proper browser and
authenticating and extracting the cookies. You will notice that in all
situations basically what happens is that we use the browser cookies so the
way you get them makes no difference on the software and I would argue that
the last method is actually the most robust since there is no automation in
between to be flaged as fraudulent by google. That being said be certain
that google invests a lot of effort to identify fraudulent traffic so I
think that in your case what happens is that your network patterns flags
your traffic as fraudulent and you get temporarily blocked. The same things
happens to me for example, whenever my wife tries to access maps from her
laptop in the house she gets a captcha prompts and I guess that my HA gets
blocked for an hour which I do not mind.

So basically there is not much we can do about it, google is constantly
tightening their fraud detection technics so if I had to guess on why it
does not work now where it did before, I would say that it is probably
because google advanced their fraud detection technics in your area and
your network traffic gets flaged. I am sorry that I cannot provide any more
that this as info. You can of course install an older version of the
library and set it up on a box and see it that works longer for you and if
that is the case then there might be something we could look into.

kind regards
Costas
